<div id="top"></div>

<!-- PROJECT LOGO -->
<br />

[//]: # (<div align="center">)

[//]: # ()
[//]: # (  <a href="https://github.com/sid-krish/rhometa">)

[//]: # ()
[//]: # (    <img src="images/logo.png" alt="Logo" width="80" height="80">)

[//]: # ()
[//]: # (  </a>)

<h1 align="center">Rhometa</h1>

  <p align="center">
    Metagenomic Population Recombination Rate Estimation Pipeline

[//]: # (    <br />)

[//]: # (    <a href="https://github.com/github_username/repo_name"><strong>Explore the docs »</strong></a>)

[//]: # (    <br />)

[//]: # (    <br />)

[//]: # (    <a href="https://github.com/github_username/repo_name">View Demo</a>)

[//]: # (    ·)

[//]: # (    <a href="https://github.com/github_username/repo_name/issues">Report Bug</a>)

[//]: # (    ·)

[//]: # (    <a href="https://github.com/github_username/repo_name/issues">Request Feature</a>)
  </p>

[//]: # (</div>)



<!-- TABLE OF CONTENTS -->
<details>
  <summary>Table of Contents</summary>
  <ol>
    <li>
      <a href="#about-the-project">About The Project</a>
      <ul>
        <li><a href="#built-with">Built With</a></li>
      </ul>
    </li>
    <li>
      <a href="#getting-started">Getting Started</a>
      <ul>
        <li><a href="#prerequisites">Prerequisites</a></li>
        <li><a href="#installation">Installation</a></li>
      </ul>
    </li>
    <li><a href="#usage">Usage</a></li>
    <li><a href="#roadmap">Roadmap</a></li>
    <li><a href="#contributing">Contributing</a></li>
    <li><a href="#license">License</a></li>
    <li><a href="#contact">Contact</a></li>
    <li><a href="#acknowledgments">Acknowledgments</a></li>
  </ol>
</details>



<!-- ABOUT THE PROJECT -->
## About The Project

[//]: # ([![Product Name Screen Shot][product-screenshot]]&#40;https://example.com&#41;)

Pipeline for estimating population recombination rates in modern metagenomic
datasets based on the LDhat composite likelihood method.
<p align="right">(<a href="#top">back to top</a>)</p>



### Built With

* [Python](https://www.python.org/)
* [Nextflow](https://www.nextflow.io/)


<p align="right">(<a href="#top">back to top</a>)</p>



<!-- GETTING STARTED -->
## Getting Started

This pipeline is designed to be run on linux and requires nextflow to be installed. 
Dependencies are resolved either via conda or docker images. Support for HPC, docker, singularity, AWS
and many other systems are provided via nextflow.

### Prerequisites
* Nextflow (required)
* conda or containerization platform (required)
  * If using conda, Conda package manager available from: [Anaconda download](https://www.anaconda.com/products/individual)
  * If using containers, docker from: [Docker download](https://www.docker.com/get-started)
  * Other container technologies such as singularity are also supported via nextflow


### Installation
Instructions for installing nextflow and dependencies via conda (default method)
1. Clone the repo
   ```sh
   git clone https://github.com/sid-krish/repo_name.git
   ```
2. Make all files in bin directory executable
   ```sh
   cd bin
   chmod +x *
   ```
3. Adjust settings in nextflow.config file, by default it is configured to work with conda with modest resources

<p align="right">(<a href="#top">back to top</a>)</p>

<!-- USAGE EXAMPLES -->
## Usage

The pipeline is made up of 4 of modules. Each of which can be used independently as necessary.
![Logo](images/modules.png)

* **sim_gen.nf** is used to generate bam files and fasta reference files with simulated reads from bacterial genomes with recombination, the bacterial genomes were simulated using fastsimbac. This module is primarily included so that the simulated datasets used for the paper can be reproduced.
  * Basic run command:
  ```sh
   cd bin
   chmod +x *
   ```
* **theta_est.nf** is used to determine the population mutation rate (theta) per site based on the LDhat method. The theta value is one the required parameters for generating lookup tables and this module helps to get an estimate on the datasets that is to be analysed. The user has the option to use the estimated theta or a different value when generating lookup tables.
* **lookup_table_gen.nf** this module makes use of LDpop and pyrho to generate the lookup tables required for the recombination rate estimator and be launched in one of 2 ways. It can either use a pre generated lookup table for high depth, which then will be downsampled for each depth from 3 to the depth of the lookup table or the pipeline can generate a high depth lookup table from scratch and then perform the downsampling step. The downsampling algorithm is a part of pyrho, it is significantly faster to generate the required smaller lookup tables from a larger table and the results are essentially identical.
* **recom_est.nf** is used to estimate the population recombination rate of metagenomic read based datasets in the form of bam and reference fasta files. It makes use of the lookup tables generated by the lookup_table_gen pipeline. The pipeline includes an option called sub_sample bam which is able to downsample the bam to the maximum specified depth, that is the depth of a largest lookup table, ensuring all the positions with a high number of alignment are not omitted from consideration. 


General nextflow help can be accessed with: 
   ```sh
    nextflow -help
   ```
\
Module specific help can be accessed with:
   ```sh
    nextflow module_name.nf --help
   ```
The module specific help also provides instructions on how to run it and the parameter options.


<p align="right">(<a href="#top">back to top</a>)</p>


<!-- CONTRIBUTING -->
## Contributing

Any contributions you make are **greatly appreciated**.

If you have a suggestion that would make this better, please fork the repo and create a pull request. You can also simply open an issue with the tag "enhancement".

<p align="right">(<a href="#top">back to top</a>)</p>



<!-- LICENSE -->
## License

Distributed under the MIT License. See `LICENSE.txt` for more information.

<p align="right">(<a href="#top">back to top</a>)</p>



<!-- CONTACT -->
## Contact

Sid Krishnan - sidaswar.krishnan-1@student.uts.edu.au

[//]: # (Project Link: [https://github.com/github_username/repo_name]&#40;https://github.com/github_username/repo_name&#41;)

<p align="right">(<a href="#top">back to top</a>)</p>



<!-- ACKNOWLEDGMENTS -->
## Acknowledgments

* []()
* []()
* []()

<p align="right">(<a href="#top">back to top</a>)</p>



<!-- MARKDOWN LINKS & IMAGES -->
<!-- https://www.markdownguide.org/basic-syntax/#reference-style-links -->
[contributors-shield]: https://img.shields.io/github/contributors/github_username/repo_name.svg?style=for-the-badge
[contributors-url]: https://github.com/github_username/repo_name/graphs/contributors
[forks-shield]: https://img.shields.io/github/forks/github_username/repo_name.svg?style=for-the-badge
[forks-url]: https://github.com/github_username/repo_name/network/members
[stars-shield]: https://img.shields.io/github/stars/github_username/repo_name.svg?style=for-the-badge
[stars-url]: https://github.com/github_username/repo_name/stargazers
[issues-shield]: https://img.shields.io/github/issues/github_username/repo_name.svg?style=for-the-badge
[issues-url]: https://github.com/github_username/repo_name/issues
[license-shield]: https://img.shields.io/github/license/github_username/repo_name.svg?style=for-the-badge
[license-url]: https://github.com/github_username/repo_name/blob/master/LICENSE.txt
[linkedin-shield]: https://img.shields.io/badge/-LinkedIn-black.svg?style=for-the-badge&logo=linkedin&colorB=555
[linkedin-url]: https://linkedin.com/in/linkedin_username
[product-screenshot]: images/screenshot.png
